### PR TITLE
MCMC implementation

### PR DESCRIPTION
**Pull request by dominikglandorf**
includes a truncated random walk proposal jointly and independently for mass and restitution

adds visualization each time proposed parameters have been accepted by MH

Screenshot 2023-09-01 at 13 42 56

another visualization summarizes all updates to the latents
Screenshot 2023-09-01 at 13 43 44

### CODE REVIEW

File: examples/mcmc.jl [hunk 1]
```diff
@@ -1,52 +1,100 @@
...
+                                low, high)
+    return Distributions.rand(d)
+end;
+function Gen.logpdf(::TruncNorm, x::Float64, mu::U, noise::T, low::T, high::T) where {U<:Real,T<:Real}
+    d = Distributions.Truncated(Distributions.Normal(mu, noise),
+                                low, high)
+    return Distributions.logpdf(d, x)
+end;
+
+# this function does not work because the max and min function do not offer an interface for the @trace macro, they would have to be implemented as probability distributions
```
Comment (belledon) [inline examples/mcmc.jl]: not super relevant but, in general, deterministic calls should not be in `@trace` expressions. anyhow, this would not serve as a good, simple proposal since it includes additional random moves that aren't present in the generative model (this would require involutive mcmc)

File: examples/mcmc.jl [hunk 2]
```diff
@@ -1,52 +1,100 @@
...
-    prev_res = TODO
+    choices = get_choices(tr)
+    prev_mass = choices[:prior => 1 => :mass]
+    prev_res  = choices[:prior => 1 => :restitution]
+    
     # sample new values conditioned on the old ones
-    mass ~ TODO
-    restitution ~ TODO
+    mass = {:prior => 1 => :mass} ~ trunc_norm(prev_mass, .1, 0., Inf)
+    restitution = {:prior => 1 => :restitution} ~ trunc_norm(prev_res, .1, 0., 1.)
```
Comment (belledon) [inline examples/mcmc.jl:33]: nice

File: examples/mcmc.jl [hunk 3]
```diff
@@ -90,17 +138,28 @@ function data_generating_procedure(t::Int64)
...
-
+    (gargs, obs, plt) = data_generating_procedure(t)
+    display(plt)
+    
+    update_vis(trace) = display(plot!(plt, 1:t, get_zs(trace), label="res $(trace[:prior => 1 => :restitution])"))
+    (tr, aratio) = inference_procedure(gargs, obs, update_vis)
+    print(tr[:prior => 1])
+
+    println("press enter to exit the program")
+    readline()
```
Comment (belledon) [inline examples/mcmc.jl:155]: why is this here?

File: examples/mcmc.jl [hunk 4]
```diff
@@ -1,52 +1,100 @@
...
+    # visualize predictions of initial trace
+    update_vis(tr)
 
     println("Initial logscore: $(ls)")
 
-    # TODO: use this to count the number of accepted moves
+    # count the number of accepted moves and track accepted proposals
     acceptance_count = 0
+    mass_log = Vector{Float64}()
+    res_log = Vector{Float64}()
```
Comment (belledon) [inline examples/mcmc.jl]: Rather than having to re-size these vectors dynamically, you can initialize them with a fixed size

```julia
v = Vector{Float64}(undef, steps) 

v[i] = value
```

see docs for [`undef`](https://docs.julialang.org/en/v1/base/arrays/#Core.undef)

File: examples/mcmc.jl [hunk 5]
```diff
@@ -1,52 +1,100 @@
...
+        push!(mass_log, tr[:prior => 1 => :mass])
+        push!(res_log, tr[:prior => 1 => :restitution])
+        acceptance_count += Int(accepted)
+        if accepted
+            update_vis(tr)
+        end
     end
+    plt = plot(1:steps, mass_log, title="Trace log of latent priors", label="mass", xlabel="step",  ylabel="value", ylim = (0, 2))
+    plot!(plt, 1:steps, res_log, label="res")
+    display(plt)
```
Comment (belledon) [inline examples/mcmc.jl]: please, return `res_log` and `mas_log`, and move the plotting to outside of `inference_procedure`. this will help with clarity - functions should be as simple as possible

File: examples/mcmc.jl [hunk 6]
```diff
@@ -90,17 +138,28 @@ function data_generating_procedure(t::Int64)
         Gen.set_submap!(obs, addr, _choices)
     end
 
-    return (gargs, obs)
+    #=  plt = lineplot(1:t, get_zs(trace),
+                   title="Height of ball", name="observation with res $(trace[:prior => 1 => :restitution])",
+                   xlabel="t", ylabel="z", canvas=DotCanvas,
+                   border=:ascii) =#
+    plt = plot(1:t, get_zs(trace),title="Height of ball", xlabel="t", ylabel="z",label="Observation")
```
Comment (belledon) [inline examples/mcmc.jl]: same as above

**Comment by belledon** [comment]
we would like to visualize a histogram of the random-walk in metropolis hastings. 

For each step (whether it is accepted or not), you should record the latents. Then, you can visualize the histogram for each latent. You could also track the log scores to plot convergence (step on x axis, and logscore on y)

File: examples/mcmc.jl (hunk 3 above)
Comment (dominikglandorf) [reply examples/mcmc.jl:155]: i added it because on my machine the plots close when the application finishes

**Comment by dominikglandorf** [comment]
I updated plotting and collected everything in main. Plots now looks like this (steps=100)

Screenshot 2023-09-04 at 15 33 04

Screenshot 2023-09-04 at 15 36 21

File: examples/mcmc.jl (hunk 3 above)
Comment (belledon) [reply examples/mcmc.jl:155]: ah, ok. i'm not the biggest fan and would prefer that you just write to disk but that's more of a style thing

File: examples/mcmc.jl [hunk 7]
```diff
@@ -137,26 +127,27 @@ function data_generating_procedure(t::Int64)
...
+
+    zs = [get_submap(obs, :kernel => i => :observe => 1)[:position][3] for i in 1:t]
+    plt = plot(1:t, zs, title="Height of ball", xlabel="t", ylabel="z", label="Observation")
     display(plt)
     
-    update_vis(trace) = display(plot!(plt, 1:t, get_zs(trace), label="res $(trace[:prior => 1 => :restitution])"))
-    (tr, aratio) = inference_procedure(gargs, obs, update_vis)
-    print(tr[:prior => 1])
+    update_vis(trace) = display(plot!(plt, 1:t, get_zs(trace), label="res $(round(trace[:prior => 1 => :restitution]; digits=3))"))
+    (tr, aratio, mass_log, res_log, scores) = inference_procedure(gargs, obs, update_vis)
```
Comment (belledon) [inline examples/mcmc.jl]: this will sound pedantic, but it will come up later when you practice with more complicated inference procedures; please remove `update_viz` from `inference_procedure`. you have several options: 
- return a vector of traces that you can then extract latents, log scores, positions etc
-  in addition to the vectors you currently return, you could also return the z positions of each trace (e.g., a nested vector of vectors, you could also used a matrix).

File: examples/mcmc.jl [hunk 8]
```diff
@@ -137,26 +127,27 @@ function data_generating_procedure(t::Int64)
...
     end
-
-    #=  plt = lineplot(1:t, get_zs(trace),
-                   title="Height of ball", name="observation with res $(trace[:prior => 1 => :restitution])",
-                   xlabel="t", ylabel="z", canvas=DotCanvas,
-                   border=:ascii) =#
-    plt = plot(1:t, get_zs(trace),title="Height of ball", xlabel="t", ylabel="z",label="Observation")
     
-    return (gargs, obs, plt)
+    return (gargs, obs)
```
Comment (belledon) [inline examples/mcmc.jl]: you could return the ground truth trace here if you wanted to access to the original set of latents. this will be good to include in your inference plots.

File: examples/mcmc.jl [hunk 9]
```diff
@@ -137,26 +127,27 @@ function data_generating_procedure(t::Int64)
...
 
 end
 
 function main()
 
     t = 60 # 1 second of observations
-    (gargs, obs, plt) = data_generating_procedure(t)
+    (gargs, obs) = data_generating_procedure(t)
+
+    zs = [get_submap(obs, :kernel => i => :observe => 1)[:position][3] for i in 1:t]
```
Comment (belledon) [inline examples/mcmc.jl]: `get_submap` is better used to collect sets of data. here you are calling it to get the positions at each time step. you could simply just use 

```julia
map(i -> obs[:kernel => i => :observe => 1], 1:t)
```

File: examples/mcmc.jl (hunk 7 above)
Comment (dominikglandorf) [reply examples/mcmc.jl]: i see. is there any other way of live monitoring during the inference procedure?

Comment (belledon) [reply examples/mcmc.jl]: what you did is legal, i just prefer minimal functions, especially for these examples. although in this case, I'm not sure what would be missing by storing the intermediate results in an array?

Comment (dominikglandorf) [reply examples/mcmc.jl]: on my computer the inference runs for two or three minutes already, this is why i liked it to see live progress and understand better what is going on over time. Of course, I can also animate this post-hoc to visualize the inference process step-wise.

Comment (belledon) [reply examples/mcmc.jl]: i can understand that. let's move on - i may make further refinements to this later
